clang/lib/Analysis/FlowSensitive/TypeErasedDataflowAnalysis.cpp

```diff
@@ -307,7 +310,22 @@ computeBlockInputState(const CFGBlock &Block, AnalysisContext &AC) {
     }
   }
 
-  JoinedStateBuilder Builder(AC);
+  // If any of the predecessor blocks contains an expression consumed in a
+  // different block, we need to keep expression state.
+  // Note that in this case, we keep expression state for all predecessors,
+  // rather than only those predecessors that actually contain an expression
+  // consumed in a different block. While this is potentially suboptimal, it's
+  // actually likely, if we have control flow within a full expression, that
+  // all predecessors have expression state consumed in a different block.
+  Environment::ExprJoinBehavior ExprBehavior = Environment::DiscardExprState;
+  for (const CFGBlock *Pred : Preds) {
+    if (Pred && AC.CFCtx.containsExprConsumedInDifferentBlock(*Pred)) {
+      ExprBehavior = Environment::KeepExprState;
+      break;
+    }
+  }
```

Comment:
Since this is a property of the CFG, can we compute this just once per block (rather than every time we visit the block)?

Reply:
I tried this out -- essentially, I added a method to `ControlFlowContext` that allows querying what the `ExprJoinBehavior` should be for a given block. However, I'm not convinced that the result was actually better that what we have now.

`ControlFlowContext` is essentially an "embellished" CFG -- it contains the CFG along with additional information derived from the CFG that the CFG itself doesn't contain. (BTW, what do you think of renaming `ControlFlowContext` to `EmbellishedCFG`? I think it describes the class better, and it would leave us with one fewer `Context`, of which we have confusingly many.)

I would argue that `exprJoinBehaviorForBlock()` isn't a direct property of the CFG block -- yes, it's a property that we associate with the block, but that property only makes sense in the context of how we join input states for the block, and as such seems more closely associated with `computeBlockInputState()` and the `Environment` than with the `ControlFlowContext`.

On the other hand, `containsExprConsumedInDifferentBlock()` _does_ express a very direct property of the block. It seems more natural to have this be the property that is exposed by `ControlFlowContext`, and to have `computeBlockInputState()` derive from this the information it needs on how to join states from the predecessor blocks.

Yes, this means we compute this information each time we visit the block, but this appears pretty insignificant compared to the work we do actually processing the block. I ran the Crubit nullability check benchmarks with the draft change I mentioned, and the speedup is too small to distinguish from measurement noise at the CPU time level, though a small significant change can be seen at the cycles level. I don't think this justifies changing the current design, which I think has better separation of concerns.

Reply:
re: EmbellishedCFG -- sure.

Ok, I'm fine leaving as is, given the benchmarks. But, I would think there's a simple compromise of just computing it locally in this file (as now), but caching it in `AnalysisContext` (which is local to that file), since its a fixed property of the block.

Reply:
I'm still not sure this is worth it - this would add additional stored state (even if it's non-mutable), and I'm not sure this is worth it for the small gain we get. Will commit as-is, and we can discuss whether we want to create a followup patch for this.